From c80424531b51663a0067a6b8ba2108f4eee9d22e Mon Sep 17 00:00:00 2001
From: Ethan Wei <ethanwei@ami.com>
Date: Mon, 16 Oct 2023 00:52:26 -0700
Subject: [PATCH 11/13] 
 0017-Fix-property-of-DomainName-in-each-EthernetInterface-Returns-Empty.patch

---
 src/dhcp_configuration.cpp                    |   1 +
 src/dns_updater.cpp                           |  26 +-
 src/ethernet_interface.cpp                    | 143 +++++--
 src/ethernet_interface.hpp                    |  13 +-
 src/network_manager.cpp                       |   2 +-
 6 files changed, 504 insertions(+), 35 deletions(-)
 create mode 100644 0017-Fix-property-of-DomainName-in-each-EthernetInterface-Returns-Empty.patch

diff --git a/src/dhcp_configuration.cpp b/src/dhcp_configuration.cpp
index 822a178..02f554e 100644
--- a/src/dhcp_configuration.cpp
+++ b/src/dhcp_configuration.cpp
@@ -55,6 +55,7 @@ Configuration::Configuration(sdbusplus::bus_t& bus,
 
     ConfigIntf::dnsEnabled(getDHCPProp(conf, "UseDNS"), true);
     ConfigIntf::ntpEnabled(getDHCPProp(conf, "UseNTP"), true);
+    ConfigIntf::domainEnabled(getDHCPProp(conf, "UseDomains"), ConfigIntf::dnsEnabled());
     ConfigIntf::hostNameEnabled(getDHCPProp(conf, "UseHostname"), true);
     ConfigIntf::sendHostNameEnabled(getDHCPProp(conf, "SendHostname"), true);
     ConfigIntf::sendNsupdateEnabled(getDHCPProp(conf, "SendNsupdate"), false);
diff --git a/src/dns_updater.cpp b/src/dns_updater.cpp
index 90644bb..f4f2638 100644
--- a/src/dns_updater.cpp
+++ b/src/dns_updater.cpp
@@ -158,20 +158,16 @@ void Configuration::registerSignal(sdbusplus::bus_t& bus) {
                     msg.read(iface, props);
                     for (const auto& t : props)
                     {
-                        if (t.first == "DomainEnabled")
+                        if (t.first == "DNSEnabled")
                         {
                             value = std::get<bool>(t.second);
+                            dnsEnabled(value);
                             if (value == true && dnsEnabled())
                                 domainConf(std::make_tuple(true,1, ""));
                             else {
                                 domainConf(std::make_tuple(false,0, std::get<2>(ddnsIface::domainConf())));
                             }
                         }
-                        else if (t.first == "DNSEnabled")
-                        {
-                            value = std::get<bool>(t.second);
-                            dnsEnabled(value);
-                        }
                         else if (t.first == "SendNsupdateEnabled")
                         {
                             value = std::get<bool>(t.second);
@@ -505,7 +501,10 @@ int16_t Configuration::setDomainConf(bool dhcp, uint8_t priority,std::string dom
 
     for (auto i = manager.get().interfaces.begin(); i != manager.get().interfaces.end(); i++) {
         if (i->second->linkUp()) {
-            i->second->domainName({});
+            if (!dhcp)
+                i->second->domainName(tmpVec);
+            else
+                i->second->domainName({});
         }
     }
 
@@ -630,7 +629,16 @@ std::string Configuration::getDomainName(std::string interface) {
         }
     }
     else {
-        return name;
+        auto iface = manager.get().interfaces.find(interface);
+        if (iface != manager.get().interfaces.end() ) {
+            if (iface->second->domainName().size() > 0) {
+                return iface->second->domainName().at(0);
+            }
+            return name;
+        }
+        else {
+            return name;
+        }
     }
 }
 
@@ -668,7 +676,7 @@ std::vector<std::string> Configuration::getDNSServer(std::string interface) {
         servers.push_back(stdplus::toStr(addrFromBuf(addrFamily, stdplus::raw::asView<char>(ipaddress))));
     }
 
-    return std::move(servers);
+    return servers;
 }
 
 
diff --git a/src/ethernet_interface.cpp b/src/ethernet_interface.cpp
index 4db9de2..c23a4f9 100644
--- a/src/ethernet_interface.cpp
+++ b/src/ethernet_interface.cpp
@@ -42,7 +42,11 @@ using NotAllowedArgument = xyz::openbmc_project::Common::NotAllowed;
 using Argument = xyz::openbmc_project::Common::InvalidArgument;
 using std::literals::string_view_literals::operator""sv;
 constexpr auto RESOLVED_SERVICE = "org.freedesktop.resolve1";
+constexpr auto RESOLVD_OBJ_PATH = "/org/freedesktop/resolve1";
 constexpr auto RESOLVED_INTERFACE = "org.freedesktop.resolve1.Link";
+constexpr auto RESOLVD_MANAGER_INTERFACE = "org.freedesktop.resolve1.Manager";
+constexpr auto DHCP_PROP_INTERFACE = "xyz.openbmc_project.Network.DHCPConfiguration";
+constexpr auto DHCP_SERVICE_PATH = "/xyz/openbmc_project/network/dhcp";
 constexpr auto PROPERTY_INTERFACE = "org.freedesktop.DBus.Properties";
 constexpr auto RESOLVED_SERVICE_PATH = "/org/freedesktop/resolve1/link/";
 
@@ -156,6 +160,8 @@ EthernetInterface::EthernetInterface(stdplus::PinnedRef<sdbusplus::bus_t> bus,
         addStaticNeigh(neigh);
     }
 
+    signals = initSignals();
+    registerSignal(bus);
     manager.get().getDNSConf().addInterfaceConf(interfaceName());
 }
 
@@ -596,10 +602,13 @@ void EthernetInterface::loadNameServers(const config::Parser& config)
         config.map.getValueStrings("Network", "DNS"));
 }
 
-void EthernetInterface::loadStaticDomainNames()
+void EthernetInterface::loadDomainNames()
 {
     auto [dhcp, priority, name] = manager.get().getDNSConf().domainConf();
-    if (!dhcp) {
+    if (manager.get().getDHCPConf().domainEnabled() && manager.get().getDHCPConf().dnsEnabled() && dhcp) {
+        EthernetInterfaceIntf::domainName(getDomainNamesFromResolvd());
+    }
+    else {
         EthernetInterfaceIntf::domainName(std::vector<std::string>{name});
     }
 }
@@ -684,6 +693,37 @@ ServerList EthernetInterface::getNameServerFromResolvd()
     return servers;
 }
 
+ServerList EthernetInterface::getDomainNamesFromResolvd()
+{
+    ServerList DomainNames;
+    auto OBJ_PATH = fmt::format("{}{}", RESOLVED_SERVICE_PATH, ifIdx);
+
+    using type = std::vector<std::tuple<std::string, bool>>;
+    std::variant<type> name; // Variable to capture the DNS property
+    auto method = bus.get().new_method_call(RESOLVED_SERVICE, OBJ_PATH.c_str(),
+                                      PROPERTY_INTERFACE, METHOD_GET);
+
+    method.append(RESOLVED_INTERFACE, "Domains");
+
+    try
+    {
+        auto reply = bus.get().call(method);
+        reply.read(name);
+    }
+    catch (const sdbusplus::exception_t& e)
+    {
+        log<level::ERR>("Failed to get DNS information from Systemd-Resolved");
+    }
+    auto tupleVector = std::get_if<type>(&name);
+    for (auto i = tupleVector->begin(); i != tupleVector->end(); ++i)
+    {
+        auto [domainName, fromRoute] = (*i);
+        DomainNames.push_back(domainName);
+    }
+
+    return DomainNames;
+}
+
 ObjectPath EthernetInterface::createVLAN(uint16_t id)
 {
     auto idStr = stdplus::toStr(id);
@@ -780,9 +820,9 @@ void EthernetInterface::writeConfigurationFile()
     }
     {
         auto& network = config.map["Network"].emplace_back();
-        auto& lla = network["LinkLocalAddressing"];
-#ifdef LINK_LOCAL_AUTOCONFIGURATION
         {
+            auto& lla = network["LinkLocalAddressing"];
+#ifdef LINK_LOCAL_AUTOCONFIGURATION
             switch (EthernetInterfaceIntf::linkLocalAutoConf()) {
                 case EthernetInterface::LinkLocalConf::v4:
                     lla.emplace_back("ipv4");
@@ -791,33 +831,37 @@ void EthernetInterface::writeConfigurationFile()
                     lla.emplace_back("no");
                     break;
                 case EthernetInterface::LinkLocalConf::v6:
-                    lla.emplace_back("ipv6");
-                    break;
                 case EthernetInterface::LinkLocalConf::both:
                 default:
-                    lla.emplace_back("yes");
+                    if (std::string{LINK_LOCAL_AUTOCONFIGURATION} == "ipv6")
+                        lla.emplace_back("ipv6");
+                    else
+                        lla.emplace_back("yes");
                     break;
             }
-        }
+
 #else
-        lla.emplace_back("no");
+            lla.emplace_back("no");
 #endif
-        config::Parser IfaceState;
-        auto& state = IfaceState.map["Network"].emplace_back();
-        state["IPv4Enable"].emplace_back(EthernetInterfaceIntf::ipv4Enable() ? "true" : "false");
-        state["IPv6Enable"].emplace_back(EthernetInterfaceIntf::ipv6Enable() ? "true" : "false");
-        IfaceState.writeFile(fs::path{fmt::format("{}/{}", manager.get().ifaceConfDir.generic_string(), interfaceName()).c_str()});
-
-        network["IPv6AcceptRA"].emplace_back(EthernetInterfaceIntf::ipv6Enable() && ipv6AcceptRA() ? "true" : "false");
-        network["DHCP"].emplace_back(dhcp4() ? (dhcp6() ? "true" : "ipv4")
-                                             : (dhcp6() ? "ipv6" : "false"));
-
-        std::string s("");
-        for (int i = 0; i < (int)domainName().size(); i++) {
-            s += domainName().at(i);
         }
-        network["Domains"].emplace_back(s);
-
+        {
+            config::Parser IfaceState;
+            auto& state = IfaceState.map["Network"].emplace_back();
+            state["IPv4Enable"].emplace_back(EthernetInterfaceIntf::ipv4Enable() ? "true" : "false");
+            state["IPv6Enable"].emplace_back(EthernetInterfaceIntf::ipv6Enable() ? "true" : "false");
+            IfaceState.writeFile(fs::path{fmt::format("{}/{}", manager.get().ifaceConfDir.generic_string(), interfaceName()).c_str()});
+
+            network["IPv6AcceptRA"].emplace_back(EthernetInterfaceIntf::ipv6Enable() && ipv6AcceptRA() ? "true" : "false");
+            network["DHCP"].emplace_back(dhcp4() ? (dhcp6() ? "true" : "ipv4")
+                                                : (dhcp6() ? "ipv6" : "false"));
+        }
+        {
+            std::string s("");
+            for (int i = 0; i < (int)domainName().size(); i++) {
+                s += domainName().at(i);
+            }
+            network["Domains"].emplace_back(s);
+        }
         {
             auto& vlans = network["VLAN"];
             for (const auto& [_, intf] : manager.get().interfaces)
@@ -1634,5 +1678,56 @@ end:
     return ret;
 }
 
+std::map<std::string, std::unique_ptr<sdbusplus::bus::match_t>> EthernetInterface::initSignals() {
+    std::map<std::string, std::unique_ptr<sdbusplus::bus::match_t>> mp;
+    mp["DHCPSignal"] = nullptr;
+    mp["ResolvdSignal"] = nullptr;
+    return mp;
+}
+
+void EthernetInterface::registerSignal(sdbusplus::bus::bus& bus) {
+    for (auto& signal : signals) {
+        if (signal.second == nullptr && signal.first == "DHCPSignal") {
+            signal.second = std::make_unique<sdbusplus::bus::match_t>(
+                bus,
+                sdbusplus::bus::match::rules::propertiesChanged(DHCP_SERVICE_PATH, DHCP_PROP_INTERFACE),
+                [&](sdbusplus::message::message& msg) {
+                    std::map<std::string, std::variant<bool>> props;
+                    std::string iface;
+                    bool value;
+                    msg.read(iface, props);
+                    for (const auto& t : props)
+                    {
+                        if (t.first == "DNSEnabled")
+                        {
+                            value = std::get<bool>(t.second);
+                            if (value) {
+                                EthernetInterfaceIntf::domainName({});
+                            }
+                        }
+                    }
+                }
+            );
+        }
+        else if (signal.second == nullptr && signal.first == "ResolvdSignal") {
+            signal.second = std::make_unique<sdbusplus::bus::match_t>(
+                bus,
+                sdbusplus::bus::match::rules::propertiesChanged(RESOLVD_OBJ_PATH, RESOLVD_MANAGER_INTERFACE),
+                [&](sdbusplus::message::message& msg) {
+                    std::map<std::string, std::variant<std::vector<std::tuple<int, std::string, bool>>,std::vector<std::tuple<int, int, std::vector<uint8_t>>>>> props;
+                    std::string iface;
+                    std::vector<std::tuple<std::string, bool>> value;
+                    msg.read(iface, props);
+                    for (const auto& t : props)
+                    {
+                        auto vector = getDomainNamesFromResolvd();
+                        EthernetInterfaceIntf::domainName(getDomainNamesFromResolvd());
+                    }
+                }
+            );
+        }
+    }
+}
+
 } // namespace network
 } // namespace phosphor
diff --git a/src/ethernet_interface.hpp b/src/ethernet_interface.hpp
index 4a27e77..ee1e9cd 100644
--- a/src/ethernet_interface.hpp
+++ b/src/ethernet_interface.hpp
@@ -126,7 +126,7 @@ class EthernetInterface : public Ifaces
 
     /** @brief Function used to load the domainNames.
      */
-    void loadStaticDomainNames();
+    void loadDomainNames();
 
     /** @brief Function to create ipAddress dbus object.
      *  @param[in] addressType - Type of ip address.
@@ -298,6 +298,15 @@ class EthernetInterface : public Ifaces
      */
     virtual ServerList getNameServerFromResolvd() const;
 
+    /** @brief get the domain names details from the network conf
+     *
+     */
+    virtual ServerList getDomainNamesFromResolvd();
+
+    std::map<std::string, std::unique_ptr<sdbusplus::bus::match_t>> initSignals();
+
+    void registerSignal(sdbusplus::bus::bus& bus);
+
     /** @brief Persistent sdbusplus DBus bus connection. */
     stdplus::PinnedRef<sdbusplus::bus_t> bus;
 
@@ -318,6 +327,8 @@ class EthernetInterface : public Ifaces
     };
     std::optional<VlanProperties> vlan;
 
+    std::map<std::string, std::unique_ptr<sdbusplus::bus::match_t>> signals ;
+
     friend class TestEthernetInterface;
     friend class TestNetworkManager;
 
diff --git a/src/network_manager.cpp b/src/network_manager.cpp
index d2a2f84..d1ac1d6 100644
--- a/src/network_manager.cpp
+++ b/src/network_manager.cpp
@@ -301,7 +301,7 @@ void Manager::createInterface(const AllIntfInfo& info, bool enabled)
         bus, *this, info, objPath.str, config, enabled);
     intf->loadNameServers(config);
     intf->loadNTPServers(config);
-    intf->loadStaticDomainNames();
+    intf->loadDomainNames();
     auto ptr = intf.get();
     interfaces.insert_or_assign(*info.intf.name, std::move(intf));
     interfacesByIdx.insert_or_assign(info.intf.idx, ptr);
-- 
2.25.1

